frontend/src/app/main/data/workspace/persistence.cljs

```diff
@@ -678,3 +682,26 @@
                (rx/buffer-until (->> frame-changes (rx/debounce 1000)))
                (rx/flat-map #(reduce set/union %))
                (rx/map #(update-frame-thumbnail %)))))))))
+
+(defn preload-data-uris
+  "Preloads the image data so it's ready when necesary"
+  []
+  (ptk/reify ::preload-data-uris
+    ptk/WatchEvent
+    (watch [_ state _]
+      (let [extract-urls
+            (fn [{:keys [metadata fill-image]}]
+              (cond
+                (some? metadata)
+                [(cfg/resolve-file-media metadata)]
+
+                (some? fill-image)
+                [(cfg/resolve-file-media fill-image)]))
+
+            uris (into #{}
+                       (comp (mapcat extract-urls)
+                             (filter some?))
```

Comment:
I think there are no need for mapcat, and everything can be simplified to `keep`. I mean, extract-urls, can return a single item without wrapping it into vector, and instead of mapcat + filter, use `keep`.

On the other hand, I think your intention was handle the case where you have the metadata and and fill-image, but in this case, the `cond` construct i think it is not adequate for this case... maybe `cond->> nil` + `cons`:

```clojure
(cond->> nil
  (some? metadata)
  (cons (cfg/resolve-file-media metadata))
  
  (some? fill-image)
  (cons (cfg/resolve-file-media fill-image)))
```

and then leave the xform as it is right now.

Reply:
I think it's better to use the mapcat thinking in the future when you could potentialy have different images in the multiple fills.